nixos/modules/system/etc/etc.nix

```diff
@@ -250,69 +250,83 @@ in
-              --options lowerdir=$tmpMetadataMount::${config.system.build.etcBasedir},${etcOverlayOptions} \
-              $tmpEtcMount
-
-            # Before moving the new /etc overlay under the old /etc, we have to
-            # move mounts on top of /etc to the new /etc mountpoint.
-            findmnt /etc --submounts --list --noheading --kernel --output TARGET | while read -r mountPoint; do
-              if [[ "$mountPoint" = "/etc" ]]; then
-                continue
-              fi
-
-              tmpMountPoint="$tmpEtcMount/''${mountPoint:5}"
-                ${
-                  if config.system.etc.overlay.mutable then
-                    ''
-                      if [[ -f "$mountPoint" ]]; then
-                        touch "$tmpMountPoint"
-                      elif [[ -d "$mountPoint" ]]; then
-                        mkdir -p "$tmpMountPoint"
-                      fi
-                    ''
-                  else
-                    ''
-                      if [[ ! -e "$tmpMountPoint" ]]; then
-                        echo "Skipping undeclared mountpoint in environment.etc: $mountPoint"
-                        continue
-                      fi
-                    ''
-                }
-              mount --bind "$mountPoint" "$tmpMountPoint"
-            done
-
-            # Move the new temporary /etc mount underneath the current /etc mount.
-            #
-            # This should eventually use util-linux to perform this move beneath,
-            # however, this functionality is not yet in util-linux. See this
-            # tracking issue: https://github.com/util-linux/util-linux/issues/2604
-            ${pkgs.move-mount-beneath}/bin/move-mount --move --beneath $tmpEtcMount /etc
+            # These don't exist yet when we're called by nixos-enter
+            mkdir --parents /.rw-etc/upper /.rw-etc/work
 
-            # Unmount the top /etc mount to atomically reveal the new mount.
-            umount --lazy --recursive /etc
+            tmpMetadataMount=$(TMPDIR="" mktemp --tmpdir=/tmp --directory -t nixos-etc-metadata.XXXXXXXXXX)
+            mount --type erofs -o ro ${config.system.build.etcMetadataImage} $tmpMetadataMount
 
-            # Unmount the temporary mount
-            umount --lazy "$tmpEtcMount"
-            rmdir "$tmpEtcMount"
+            # There was no previous /etc mounted. This happens when we're called by nixos-enter
+            if ! mountpoint -q /etc; then
+              mount --type overlay overlay \
+                --options lowerdir=$tmpMetadataMount::${config.system.build.etcBasedir},${etcOverlayOptions} \
+                /etc
+            else
+              # Mount the new /etc overlay to a temporary private mount.
+              # This needs the indirection via a private bind mount because you
+              # cannot move shared mounts.
+              tmpEtcMount=$(TMPDIR="" mktemp --tmpdir=/tmp --directory -t nixos-etc.XXXXXXXXXX)
```

Comment:
Here and above we force `TMPDIR` to empty to avoid an issue with this variable leaking into the chroot from outside when using nixos-install.
I didn't find a better way to fix this yet.

In any case, further down in this script we currently also assume that `TMPDIR=/tmp`, so we're not losing much currently.

Reply:
This was changed to use /run in a later commit.